syntax = "proto3";

package multitenant.v1;

import "common.proto";

option cc_enable_arenas = true;

// User represents a user within a tenant
message User {
    int64 id = 1;
    string username = 2;
    string email = 3;
    string role = 4;
    bool active = 5;
    string created_at = 6;
    string updated_at = 7;
}

// Permission represents an access control entry
message Permission {
    int64 id = 1;
    int64 user_id = 2;
    string resource = 3;
    string action = 4;
    bool allowed = 5;
}

// UserService manages users within a tenant context
// Note: tenant_id comes from request metadata (x-tenant-id header)
service UserService {
    // Get a single user by ID
    rpc GetUser(GetUserRequest) returns (GetUserResponse);
    
    // Get user by username
    rpc GetUserByUsername(GetUserByUsernameRequest) returns (GetUserResponse);
    
    // List all users in the tenant
    rpc ListUsers(ListUsersRequest) returns (ListUsersResponse);
    
    // Create a new user
    rpc CreateUser(CreateUserRequest) returns (CreateUserResponse);
    
    // Update an existing user
    rpc UpdateUser(UpdateUserRequest) returns (UpdateUserResponse);
    
    // Delete (deactivate) a user
    rpc DeleteUser(DeleteUserRequest) returns (DeleteUserResponse);
    
    // Authentication
    rpc Authenticate(AuthenticateRequest) returns (AuthenticateResponse);
    
    // Permission management
    rpc GetUserPermissions(GetUserPermissionsRequest) returns (GetUserPermissionsResponse);
    rpc GrantPermission(GrantPermissionRequest) returns (GrantPermissionResponse);
    rpc RevokePermission(RevokePermissionRequest) returns (RevokePermissionResponse);
    rpc CheckPermission(CheckPermissionRequest) returns (CheckPermissionResponse);
}

// GetUser
message GetUserRequest {
    int64 user_id = 1;
}

message GetUserByUsernameRequest {
    string username = 1;
}

message GetUserResponse {
    User user = 1;
}

// ListUsers
message ListUsersRequest {
    Pagination pagination = 1;
    bool active_only = 2;
    string role_filter = 3;
}

message ListUsersResponse {
    repeated User users = 1;
    PaginationInfo pagination = 2;
}

// CreateUser
message CreateUserRequest {
    string username = 1;
    string email = 2;
    string password = 3;
    string role = 4;
}

message CreateUserResponse {
    User user = 1;
}

// UpdateUser
message UpdateUserRequest {
    int64 user_id = 1;
    optional string username = 2;
    optional string email = 3;
    optional string password = 4;
    optional string role = 5;
    optional bool active = 6;
}

message UpdateUserResponse {
    User user = 1;
}

// DeleteUser
message DeleteUserRequest {
    int64 user_id = 1;
    bool permanent = 2;
}

message DeleteUserResponse {
    bool success = 1;
}

// Authenticate
message AuthenticateRequest {
    string username = 1;
    string password = 2;
}

message AuthenticateResponse {
    bool success = 1;
    User user = 2;
    string token = 3;        // JWT or session token
    int64 expires_at = 4;    // Unix timestamp
}

// GetUserPermissions
message GetUserPermissionsRequest {
    int64 user_id = 1;
}

message GetUserPermissionsResponse {
    repeated Permission permissions = 1;
}

// GrantPermission
message GrantPermissionRequest {
    int64 user_id = 1;
    string resource = 2;
    string action = 3;
}

message GrantPermissionResponse {
    Permission permission = 1;
}

// RevokePermission
message RevokePermissionRequest {
    int64 user_id = 1;
    string resource = 2;
    string action = 3;
}

message RevokePermissionResponse {
    bool success = 1;
}

// CheckPermission
message CheckPermissionRequest {
    int64 user_id = 1;
    string resource = 2;
    string action = 3;
}

message CheckPermissionResponse {
    bool allowed = 1;
}
